src/libcore/str/mod.rs

```diff
@@ -1538,6 +1548,9 @@ fn run_utf8_validation(v: &[u8]) -> Result<(), Utf8Error> {
             if align != usize::max_value() && align.wrapping_sub(index) % usize_bytes == 0 {
                 let ptr = v.as_ptr();
                 while index < blocks_end {
+                    // SAFETY: since `align - index` and `ascii_block_size` are multiples of
+                    // `usize_bytes`, `ptr.add(index)` is always aligned with a `usize` so we
+                    // may cast directly to a `const` pointer.
```

Comment:
Casting to a raw const ptr is always possible so I a somewhat puzzled by this comment. Also, why does this implication hold? Not sure how you are going from `align-index` to `ptr.add(index)´.

And it's not really  `align-index` anyway, it's `wrapping_sub`, but `add` must *not* overflow. So why does this all fit together?^^ (Probably this should be reviewed by someone who has seen this code before...)

Reply:
Yeah, this is one part where I'm still unsure that it's correct, even after reading through the code a dozen times. I'd appreciate hearing why this works from someone who's familiar with the code, too.

Reply:
This took me a while to understand, but basically:
- `align` is an offset in the string at which point `ptr.add(align)` is guaranteed to be usize-aligned.
- If `align - index` is a multiple of `usize_bytes` then `ptr.add(index)` is usize-aligned.